### PR TITLE
KAFKA-14327: Unify KRaft snapshot generation between broker and controller

### PR DESCRIPTION
**Pull request by cmccabe**
Unify KRaft metadata snapshot generation between the broker and the controller, so that it all flows through the MetadataImage#write code path.

Previously, the broker generated MetadataImage objects inside BrokerMetadataListener, and published them to a single publisher, named BrokerMetadataPublisher. Now, with this PR, MetadataImage objects are loaded by the MetadataLoader, a generic metadata class shared between controller and broker. The image objects are now published to a list of publishers, rather than just to a single publisher.  Also, publishers can also be dynamically added and removed. On the broker side, we use the dynamic addition and removal to only add BrokerMetadataPublisher when our metadata has "caught up" to the controller's.

With this PR, snapshot generation for both the controller and the broker is handled by a metadata publisher named SnapshotGenerator. This replaces BrokerMetadataSnapshotter and the controller's SnapshotGenerator. This means that the code for generating snapshots from the controller data structures can be deleted and we can maintain only one code path here. The new SnapshotGenerator implements time-based snapshotting as specified by KIP-876.

Using the aforementioned image publisher framework, this PR adds support for dynamic configuration
to the controller. Currently, only socket server, authorizer, and metrics reporter reconfiguration
is supported, but more will be added in the future.

This PR also changes how we handle changing the metadata version in MetadataDelta and MetadataImage. Previously, we called handleMetadataVersionChange on all delta objects whenever we encountered a metadata change in the log. This approach was difficult to maintain, however, since changing metadata versions can add or remove fields from both deltas and images. In practice, we did not correctly implement version changes, which is why we had to disable KRaft downgrades in the 3.3 release. The new approach here is to perform an image save and load cycle. That way we only have to maintain one code path for upgrades and downgrades, in MetadataImage#write. Because doing a save and load cycle is somewhat out of scope for MetadataDelta itself, it simply throws a MetadataVersionChangedException and lets the MetadataLoader handle the change.

On a related note, this PR removes offset and epoch tracking from MetadataDelta. This tracking is done by external classes (such as MetadataLoader) so it doesn't need to be done in MetadataDelta. It also introduces a builder for MetadataDelta so that we can add new constructor parameters more easily in the future. (The only new parameter here is MetadataVersion.) Finally, there is a new class, MetadataProvenance, which encompasses offset, epoch, and timestamp, which identifies the source (provenance) of all MetadataImages.

### CODE REVIEW

File: metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java [hunk 1]
```diff
@@ -467,36 +467,41 @@ public void testIterator(MetadataVersion metadataVersion) throws Exception {
...
+        assertEquals(new ApiMessageAndVersion(new RegisterBrokerRecord().
                 setBrokerEpoch(100).setBrokerId(2).setRack(null).
                 setEndPoints(new BrokerEndpointCollection(Collections.singleton(
                     new BrokerEndpoint().setSecurityProtocol(SecurityProtocol.PLAINTEXT.id).
                         setPort((short) 9094).
                         setName("PLAINTEXT").
                         setHost("example.com")).iterator())).
-                setFenced(true), expectedVersion))),
-                clusterControl.iterator(Long.MAX_VALUE));
+                    setFenced(true), expectedVersion),
```
Comment (hachikuji) [inline metadata/src/test/java/org/apache/kafka/controller/ClusterControlManagerTest.java]: nit: looks misaligned

File: server-common/src/test/java/org/apache/kafka/queue/KafkaEventQueueTest.java [hunk 2]
```diff
@@ -240,4 +242,29 @@ public void handleException(Throwable e) {
...
-}
+
+    @Test
+    public void testEmpty() throws Exception {
+        KafkaEventQueue queue = new KafkaEventQueue(Time.SYSTEM, new LogContext(),
+                "testEmpty");
+        assertTrue(queue.isEmpty());
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        queue.append(() -> future.get());
+        assertFalse(queue.isEmpty());
```
Comment (hachikuji) [inline server-common/src/test/java/org/apache/kafka/queue/KafkaEventQueueTest.java:253]: This assertion is flaky locally. Can we use `waitForCondition` for all cases where we expect the queue to become empty?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+            }
+            metrics.updateBatchSize(batch.records().size());
+            lastOffset = batch.lastOffset();
+            numBytes += batch.sizeInBytes();
+            numBatches++;
+        }
+        MetadataProvenance provenance =
+                new MetadataProvenance(lastOffset, lastEpoch, lastContainedLogTimeMs);
+        long elapsedNs = time.nanoseconds() - startNs;
+        // TODO: this metric should be renamed something like "delta processing time"
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: nit: can we move this comment to a jira?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 4]
```diff
@@ -726,6 +728,7 @@ object KafkaConfig {
   val MetadataLogDirDoc = "This configuration determines where we put the metadata log for clusters in KRaft mode. " +
     "If it is not set, the metadata log is placed in the first log directory from log.dirs."
   val MetadataSnapshotMaxNewRecordBytesDoc = "This is the maximum number of bytes in the log between the latest snapshot and the high-watermark needed before generating a new snapshot."
+  val MetadataSnapshotMaxIntervalMsDoc = "This is the maximum number of milliseconds that can elapse between the previous metadata snapshot and the next one."
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Perhaps add something like this?

> In other words, metadata snapshots will be generated at least as frequently as specified by this configuration. Note that if `metadata.log.max.record.bytes.between.snapshots` bytes of new data is reached before the full specified interval completes, then the snapshot will still be generated.

File: core/src/main/scala/kafka/server/DynamicBrokerConfig.scala [hunk 5]
```diff
@@ -269,6 +270,26 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
...
+      case _ =>
+    }
+    if (!kafkaConfig.processRoles.contains(BrokerRole)) {
+      // In combined mode, these elements are shared between broker and controller, and therefore
+      // don't need to be updated here (since the broker will do it.)
+      addReconfigurable(controller.kafkaYammerMetrics)
+      addReconfigurable(new DynamicMetricsReporters(
+        kafkaConfig.brokerId, controller.config, controller.metrics, controller.clusterId))
+    }
+    // TODO: add dynamic thread pool resizing here
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/DynamicBrokerConfig.scala]: nit: create jiras. We need one for the quota callback as well.

File: core/src/main/scala/kafka/server/ControllerServer.scala [hunk 6]
```diff
@@ -258,6 +264,14 @@ class ControllerServer(
...
         throw e
     }
+    config.dynamicConfig.addReconfigurables(this)
+    dynamicConfigHandlers = immutable.Map[String, ConfigHandler](
+      ConfigType.Broker -> new BrokerConfigHandler(config, quotaManagers))
+    controllerMetadataPublisher = new ControllerMetadataPublisher(
+      config,
+      dynamicConfigHandlers,
+      fatalFaultHandler)
+    metadataLoader.installPublishers(List().asJava)
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/ControllerServer.scala]: Don't we need the publisher in here? Do we have any integration tests which verify this all works?

File: core/src/test/java/kafka/testkit/KafkaClusterTestKit.java [hunk 7]
```diff
@@ -191,21 +191,36 @@ public KafkaClusterTestKit build() throws Exception {
...
+                                KafkaRaftServer.configSchema(),
+                                raftManager.apiVersions(),
+                                bootstrapMetadata,
+                                metadataFaultHandler,
+                                fatalFaultHandler,
+                                metadataLoader);
+                    } catch (Throwable e) {
+                        log.error("Error creating controller {}", node.id(), e);
+                        metadataLoader.close();
+                        controller.shutdown();
```
Comment (hachikuji) [inline core/src/test/java/kafka/testkit/KafkaClusterTestKit.java]: There should probably be a null check. Same for `broker` below.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+
+    /**
+     * The publishers which should receive cluster metadata updates.
+     */
+    private final List<MetadataPublisher> publishers;
+
+    /**
+     * The current leader and epoch.
+     */
+    private LeaderAndEpoch currentLeaderAndEpoch = LeaderAndEpoch.UNKNOWN;
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:158]: This field is not accessed. Do we need it?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+                    handleMetadataVersionChange(delta,
+                        e.change(),
+                        batch.baseOffset() + indexWithinBatch,
+                        lastEpoch,
+                        lastContainedLogTimeMs);
+                } catch (Throwable e) {
+                    faultHandler.handleFault("Error loading metadata log record from offset " +
+                            batch.baseOffset() + indexWithinBatch, e);
+                }
+                lastEpoch = batch.epoch();
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: nit: pull these two lines outside the loop since they do not depend on the record

File: metadata/src/main/java/org/apache/kafka/image/loader/LogDeltaManifest.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+public class LogDeltaManifest {
+    /**
+     * The highest offset and epoch included in this delta, inclusive.
+     */
+    private final MetadataProvenance provenance;
+
+    /**
+     * The number of batches that were loaded.
+     */
+    private final int numBatches;
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/LogDeltaManifest.java:37]: Seems we didn't need this and `elapsedNs` below?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+            MetadataDelta delta,
+            SnapshotReader<ApiMessageAndVersion> reader
+    ) {
+        long startNs = time.nanoseconds();
+        int snapshotIndex = 0;
+        while (reader.hasNext()) {
+            Batch<ApiMessageAndVersion> batch = reader.next();
+            for (ApiMessageAndVersion record : batch.records()) {
+                try {
+                    delta.replay(record.message());
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:403]: This may still throw `MetadataVersionChangeException`, but we are not catching it. Am I missing something?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 12]
```diff
@@ -0,0 +1,504 @@
...
+     *
+     * @param newPublishers     The publishers to install.
+     *
+     * @return                  A future which yields null when the publishers have been added, or
+     *                          an exception if the installation failed.
+     */
+    public CompletableFuture<Void> installPublishers(List<? extends MetadataPublisher> newPublishers) {
+        if (newPublishers.isEmpty()) return CompletableFuture.completedFuture(null);
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        eventQueue.beginShutdown("installPublishers", () -> {
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: Should be `append`, right? Same for `removeAndClosePublisher`. Do we have tests that are failing?

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+            image.write(writer, new ImageWriterOptions.Builder().
+                    setMetadataVersion(image.features().metadataVersion()).
+                    build());
+            writer.close(true);
+        } catch (Throwable e) {
+            log.error("Encountered error while writing {}", provenance.snapshotName(), e);
+            throw e;
+        } finally {
+            try {
+                snapshotWriter.get().close();
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java]: This would be clearer without using the try-with-resources, which would also end up closing the writer. I do not know which logic will execute first. Perhaps it's best not to depend on the order.

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            RaftClient<ApiMessageAndVersion> raftClient,
+            int batchSize
+    ) {
+        this.log = new LogContext("[SnapshotEmitter id=" + nodeId + "] ").logger(SnapshotEmitter.class);
+        this.raftClient = raftClient;
+        this.batchSize = batchSize;
+    }
+
+    @Override
+    public void emit(MetadataImage image) {
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java]: nit: maybe we could call this `maybeEmit` or return a boolean or something. Just to make it clear that a snapshot may or may not be emitted by the call.

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java [hunk 15]
```diff
@@ -0,0 +1,262 @@
...
+            return this;
+        }
+
+        public Builder setMinTimeSinceLastSnapshotNs(long minTimeSinceLastSnapshotNs) {
+            this.minTimeSinceLastSnapshotNs = minTimeSinceLastSnapshotNs;
+            return this;
+        }
+
+        public SnapshotGenerator build() {
+            if (emitter == null) throw new RuntimeException("You must set a snapshot emitter.");
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: nit: if this is required, perhaps we should pass it in the constructor

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+        MetadataImage newImage,
+        LogDeltaManifest manifest
+    ) {
+        bytesSinceLastSnapshot += manifest.numBytes();
+        if (bytesSinceLastSnapshot >= minBytesSinceLastSnapshot) {
+            if (eventQueue.isEmpty()) {
+                scheduleEmit("we have replayed at least " + minBytesSinceLastSnapshot +
+                    " bytes", newImage);
+            }
+        } else if (minTimeSinceLastSnapshotNs != 0 &&
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: I thought the intent was to disable time-based snapshots with -1.

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+        this.nodeId = nodeId;
+        this.time = time;
+        this.emitter = emitter;
+        this.faultHandler = faultHandler;
+        this.minBytesSinceLastSnapshot = minBytesSinceLastSnapshot;
+        this.minTimeSinceLastSnapshotNs = minTimeSinceLastSnapshotNs;
+        LogContext logContext = new LogContext("[SnapshotGenerator " + nodeId + "] ");
+        this.log = logContext.logger(SnapshotGenerator.class);
+        this.disabledReason = null;
+        this.eventQueue = new KafkaEventQueue(time, logContext, "");
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: Perhaps we should add a thread name prefix?

File: core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala [hunk 18]
```diff
@@ -1,413 +0,0 @@
...
-      generateManyRecords(listener, 5000L)
-      assertEquals(5000L, snapshotter.activeSnapshotOffset)
-      assertEquals(null, snapshotter.failure.get())
-    } finally {
-      listener.close()
-    }
-  }
-
-  @Test
-  def testNotSnapshotAfterMetadataVersionChangeBeforePublishing(): Unit = {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala:280]: Are all of these tests recreated somewhere?

File: metadata/src/main/java/org/apache/kafka/image/loader/SnapshotManifest.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    ) {
+        this.provenance = provenance;
+        this.elapsedNs = elapsedNs;
+    }
+
+    public MetadataProvenance provenance() {
+        return provenance;
+    }
+
+    public long elapsedNs() {
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/SnapshotManifest.java:51]: I don't think this is used. Can we remove it?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 20]
```diff
@@ -0,0 +1,504 @@
...
+     * @param preChangeEpoch    The log epoch BEFORE the change.
+     */
+    void handleMetadataVersionChange(
+        MetadataDelta delta,
+        MetadataVersionChange change,
+        long changeOffset,
+        int preChangeEpoch,
+        long lastContainedLogTimeMs
+    ) {
+        // First, we materialize the current metadata image and send it to all the publishers that
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:314]: It would be helpful to add some motivation for this. Is it for recovery if something goes wrong after upgrade/downgrade?

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java [hunk 21]
```diff
@@ -0,0 +1,504 @@
...
+     * @param delta     The metadata delta we are preparing.
+     * @param reader    The reader which yields the snapshot batches.
+     * @return          A manifest of what was loaded.
+     */
+    SnapshotManifest loadSnapshot(
+            MetadataDelta delta,
+            SnapshotReader<ApiMessageAndVersion> reader
+    ) {
+        long startNs = time.nanoseconds();
+        int snapshotIndex = 0;
```
Comment (hachikuji) [inline metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:398]: Not sure how useful this index is given the presence of control records in the snapshot. Why not use the offset from `Batch` as we do when processing log records?

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java [hunk 22]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    public void beginShutdown() {
+        this.disabledReason = "we are shutting down";
+        eventQueue.beginShutdown("beginShutdown");
+    }
+
+    public void disable(String disabledReason) {
+        log.error("Disabling periodic KRaft snapshot generation because {}.", disabledReason);
+        this.disabledReason = disabledReason;
```
Comment (mumrah) [inline metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: I think if we pass `null` here, it will re-enable snapshot generation. Do we want to allow that? Or should snapshot disabling be irreversible until restart?

File: core/src/main/scala/kafka/server/KafkaRaftServer.scala [hunk 23]
```diff
@@ -81,26 +83,69 @@ class KafkaRaftServer(
     controllerQuorumVotersFuture
   )
 
+  private val snapshotEmitter: SnapshotEmitter = new SnapshotEmitter.Builder().
+    setNodeId(config.nodeId).
+    setRaftClient(raftManager.client).
+    build()
+
+  private val snapshotGenerator: SnapshotGenerator = new SnapshotGenerator.Builder().
+    setTime(Time.SYSTEM).
```
Comment (mumrah) [inline core/src/main/scala/kafka/server/KafkaRaftServer.scala]: Should be the `time: Time` passed into KafkaRaftServer

File: core/src/main/scala/kafka/server/KafkaRaftServer.scala [hunk 24]
```diff
@@ -81,26 +83,69 @@ class KafkaRaftServer(
...
+
+  private val brokerMetrics: Option[BrokerServerMetrics] = if (config.processRoles.contains(BrokerRole)) {
+    Some(BrokerServerMetrics(metrics))
+  } else {
+    None
+  }
+
+  private val metadataLoader: MetadataLoader = {
+    val builder = new MetadataLoader.Builder().
+      setTime(Time.SYSTEM).
```
Comment (mumrah) [inline core/src/main/scala/kafka/server/KafkaRaftServer.scala]: Same comment as above ^

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 3 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: OK. I filed [KAFKA-14348](https://issues.apache.org/jira/browse/KAFKA-14348) for this.

I will also remove this comment ... it's probably a distraction having it here

File: core/src/main/scala/kafka/server/DynamicBrokerConfig.scala (hunk 5 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/DynamicBrokerConfig.scala]: Good point. I created KAFKA-14349, KAFKA-14350, and KAFKA-14351. I will also add comments about them here.

KAFKA-14351 is for controller mutation quotas as a whole (I thought we had that already but I didn't see it, so I created it and linked the appropriate places.)

File: core/src/main/scala/kafka/server/ControllerServer.scala (hunk 6 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/ControllerServer.scala]: Yes, good catch. I am adding more tests for this (maybe not tonight but tomorrow morning for sure)

File: core/src/test/java/kafka/testkit/KafkaClusterTestKit.java (hunk 7 above)
Comment (cmccabe) [reply core/src/test/java/kafka/testkit/KafkaClusterTestKit.java]: good catch

File: metadata/src/main/java/org/apache/kafka/image/loader/LogDeltaManifest.java (hunk 10 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/LogDeltaManifest.java:37]: Sorry, this information was supposed to be logged at DEBUG level like in the old loader, but I forgot. I'll add it.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 11 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:403]: No, this was a bug. Good catch (btw, I fixed it today while testing some unrelated things)

The scenario where you might hit this is if you had an existing image in memory, and you loaded a snapshot for whatever reason, and the snapshot changed the MV.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 12 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: I have now added a MetadataLoaderTest which helped catch and fix a lot of these things.

I forgot to add it the first time around (I had one in some of the draft PRs, but while preparing the final PR I forgot to port it over) :X

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java (hunk 13 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java]: Yeah, that's a fair point. I'm sure that this ordering is defined by Java, but it's a bit esoteric.

I was assuming that the close order didn't matter, but perhaps it does (or will eventually?) Best just to be straightforward about it. Since they're both AutoCloseable, I will also use `Utils.closeQuietly` as well.

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java (hunk 14 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotEmitter.java]: Fair enough. I will also add some JavaDoc to `SnapshotGenerator.Emitter`

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 9 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java]: This was originally so that we could get the pre-change "provenance" (offset, epoch, timestamp) for writing the snapshot-on-mv-change, if that was needed.

However, I think I found a better way to do this that doesn't require updating those variables on every loop iteration.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 8 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:158]: it would be nice to keep to minimize divergence (although I don't feel strongly)

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java (hunk 15 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: Typically I don't add mandatory params to builder constructors because part of the"value add" builders provide is a visually nicer way of identifying (essentially) constructor parameters. Basically, since we don't have named parameters in Java, we can sort of fake it this way. A mandatory parameter may eventually also become optional, over time.

We wouldn't want to have builders for every object since there is some overhead, both in runtime and visual terms. But for big fat objects (which you typically don't have a lot of anyway), it really does make life a lot nicer.

[ Side note: I really regret not creating builders on day 1 for BrokerServer and ControllerServer since changing the constructor parameters is now an exercise in suffering, even with Scala's named parameters. But I never seem to find a good time to do that refactor. ]

Anyway, in this case, I'll add it, since a SnapshotGenerator without an emitter would be silly.

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java (hunk 16 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: Hmm [KIP-876](https://cwiki.apache.org/confluence/display/KAFKA/KIP-876%3A+Time+based+cluster+metadata+snapshots) currently says 0 means disabled ...

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java (hunk 17 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: fair enough. we have all the information here, so there is no reason why not (and some good reasons why to!)

File: core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala (hunk 18 above)
Comment (cmccabe) [reply core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataListenerTest.scala:280]: in the latest PR, `MetadataLoaderTest.java` replaces `BrokerMetadataListenerTest.scala`

File: metadata/src/main/java/org/apache/kafka/image/loader/SnapshotManifest.java (hunk 19 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/SnapshotManifest.java:51]: similarly to the log delta one, we need it for DEBUG log4j logging (which was missing, but I will add)

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 20 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:314]: The issue is that different publishers may be installed at different times.

For example, let's say you have some hypothetical cluster balancing service that can be dynamically enabled. (not included in this PR, of course). When it's dynamically enabled, it registers as a metadata publisher, to get all the updates.

If you enable this thing well after startup, you will need to send it a MetadataDelta that contains all the updates (since all the updates are new to it, although not to the existing publishers). How do you get such a MetadataDelta? Write out the existing image to a delta.

Basically this PR adds dynamic installation / removal to the system, which it didn't previously have. That's why the old code locks it to a single publisher. But that isn't very dynamic and we have some use cases for installing things at runtime.

File: metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java (hunk 21 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/loader/MetadataLoader.java:398]: Hmm. I'm actually a bit confused about what that field in the batch contains, since from the KRaft point of view all the records in a snapshot are at the same "offset"

Since snapshotIndex is ported from the old code, perhaps we can look into it a bit later? Or do a follow-on JIRA...

File: metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java (hunk 22 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/image/publisher/SnapshotGenerator.java]: Yeah, I think we should not allow re-enablement. I will add a `requireNonNull`

File: server-common/src/test/java/org/apache/kafka/queue/KafkaEventQueueTest.java (hunk 2 above)
Comment (cmccabe) [reply server-common/src/test/java/org/apache/kafka/queue/KafkaEventQueueTest.java:253]: [ I didn't forget this comment -- let me look into it tomorrow morning ]

Comment (cmccabe) [reply server-common/src/test/java/org/apache/kafka/queue/KafkaEventQueueTest.java:253]: This had a bug, which is that it wasn't taking into account the currently running event in its calculation of "empty". I fixed this bug and the test should be working now. I don't think we need the waits (except in the one position)

File: core/src/main/scala/kafka/server/KafkaRaftServer.scala [hunk 25]
```diff
@@ -69,95 +64,54 @@ class KafkaRaftServer(
...
     controller.foreach(_.startup())
     broker.foreach(_.startup())
     AppInfoParser.registerAppInfo(Server.MetricsPrefix, config.brokerId.toString, metrics, time.milliseconds())
     info(KafkaBroker.STARTED_MESSAGE)
   }
 
   override def shutdown(): Unit = {
+    // The last component to be shut down will stop JointServer.
     broker.foreach(_.shutdown())
-    // The order of shutdown for `RaftManager` and `ControllerServer` is backwards
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/KafkaRaftServer.scala:146]: It looks like we lost this in the updated implementation? The socket server of the controller is shutdown before `JointServer`. By the time we get to `RaftManager` shutdown, the socket server is down and we cannot do graceful resignation from the quorum anymore.

File: core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+import kafka.server.{ConfigEntityName, ConfigHandler, ConfigType, KafkaConfig}
+import kafka.utils.Logging
+import org.apache.kafka.common.config.ConfigResource.Type.{BROKER, TOPIC}
+import org.apache.kafka.image.loader.{LogDeltaManifest, SnapshotManifest}
+import org.apache.kafka.image.publisher.MetadataPublisher
+import org.apache.kafka.image.{MetadataDelta, MetadataImage}
+import org.apache.kafka.server.fault.FaultHandler
+
+
+class DynamicConfigPublisher(
```
Comment (hachikuji) [inline core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala:31]: Can we pull this into a separate PR? It is asking a lot from reviewers to be rigorous with a 4k diff. As I see it, there are at least 3 separate PRs here:

1. General refactor of server startup/shutdown
2. Implementation of time-based snapshots
3. Implementation of dynamic controller configuration

Comment (cmccabe) [reply core/src/main/scala/kafka/server/metadata/DynamicConfigPublisher.scala:31]: Here's a PR for just JointServer and nothing else : https://github.com/apache/kafka/pull/12837

I would do split 2 and  split 3 differently, as "new code" versus "changes to old code." Let me know if that makes sense.

**Comment by cmccabe** [comment]
The controller-side changes are now here: #12983

**Comment by cmccabe** [comment]
I split this PR into a lot of PRs and they are now all in. So time to close this!

Thanks, all.
